### ISSUE TITLE
[UX] Add a "Add new style" link next to the "Preview image style" drop-down menu in the image field settings form.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1438]
Once again, a "put things where people expect to find them" or "connect the dots" thing:

![backdrop-add_image_style_link-connect_the_dots](https://cloud.githubusercontent.com/assets/2423362/11923430/6f7f51de-a7f7-11e5-85e4-6f88c4a07d43.png)

Brought up by @serundeputy in https://github.com/backdrop/backdrop-issues/issues/1423#issuecomment-166159728

**Comment by jenlampton** [comment]
love it. :+1:

**Comment by docwilmot** [comment]
Don't know about this. Unless this goes in a modal, clicking that link would cause the user to lose all form fields they've already filled. Especially down there at the end of the form, thats likely to tick people off.

And its quite a mutistep form, so modal, no. 

I'd not do this.

**Comment by jenlampton** [comment]
No work done on this, bumping to 1.4.0

**Comment by serundeputy** [comment]
@docwilmot I agree;

@klonos and @docwilmot this is what I was thinking:

![link-to-img-style-if-desired-sans](https://cloud.githubusercontent.com/assets/6763671/12341614/f0a6deba-baf0-11e5-8017-afb691059bef.png)

and

![img-style-for-node](https://cloud.githubusercontent.com/assets/6763671/12341621/f82d9b2e-baf0-11e5-81fb-66eccb9d0e14.png)

does that make sense?

**Comment by klonos** [comment]
@docwilmot :

> ...clicking that link would cause the user to lose all form fields they've already filled. ...

Would suggesting implementing temp store for field setting forms (or any form in general) be overkill? If we had that, then we could either use a modal or a proper page and redirect back to the form once done. I'm thinking modal better though so to keep some sort of context.

> And the add new style form is quite a mutistep form, so modal, no.

We have multi-step forms in modals in views and it all works fine. Right?

**Comment by docwilmot** [comment]
> We have multi-step forms in modals in views and it all works fine. Right?

My thought was that modals tend to be launched as part of code also dealing with the underlying page. This situation is different though; we'd be opening an image modal on a node or field page. This I imagine would require refactoring the Image module code as well as Node/and/or Field to make it work.

But I'm just guessing, so perhaps we should await an expert opinion.

**Comment by quicksketch** [comment]
I agree with @docwilmot. Putting the entire configuration of an image style in a dialog would be challenging. There are just too many steps involved in making an image style. The only way I can think of making this possible would be if we first made the configuration of image styles an entirely one-page process (with lots of AJAX). Then once that was done, this one-page/AJAX-ified configuration of an image style could be put in a dialog. Even then, configuring an image style can be a long process. I'm not sure it should be done in a dialog.

If we wanted to just navigate the user to the image style settings page, that would be a much simpler fix, but it would introduce a different UX issue (either opening a new window or wiping out the form, as noted already).

**Comment by docwilmot** [comment]
I think we can reasonably say this is a no-go.

**Comment by klonos** [comment]
I realize there are technical implications to getting something like that working and from what you say it might not be a trivial task, but there is some really bad UX here...

User starts adding an image field. They fill in the form and at some point they reach the point where they are asked to pick a preview style only to find that the default ones do not suit their case. Back to the drawing board: they open a new tab and create an image style separately. Once they have the image style(s) they need, they cannot simply continue from where they left off because for the new styles to become available as options in the drop-down, they need to refresh the page. Oops!, you need to re-enter all data in the form you were in the middle of. Sorry.

We can do better than that I'm sure. So, the solution I proposed might not be easy or the right one to go with, but the UX issue remains and we need to figure a solution. File a separate issue?

**Comment by mikemccaffrey** [comment]
> User starts adding an image field. They fill in the form and at some point they reach the point where they are asked to pick a preview style only to find that the default ones do not suit their case

To me, this does not seem like a workflow issue that would effect most users, and only once or twice for those it does. I've only rarely set more than one preview style for add/edit forms. I usually just edit the existing preview style to meet the majority of my needs and then maybe create one more for previewing banner images.

Does anyone have an example of a site where they've created a bunch if different preview styles?

**Comment by Graham-72** [comment]
I do find myself adding image styles to a site as the design progresses and then having to rationise later. I should probably do more planning at the start. Here are a couple of examples of the result (so far):
![image styles - browsers bookshop 2016-03-29 09-37-55](https://cloud.githubusercontent.com/assets/10559274/14102518/fcf021b8-f592-11e5-8a6d-4c3832ec30e2.png)
![image styles - deben rowing club 2016-03-29 09-37-20](https://cloud.githubusercontent.com/assets/10559274/14102521/0490493e-f593-11e5-96db-b4bddd14c709.png)

**Comment by klonos** [comment]
I think that I work more like @Graham-72 too and so I get to this situation a bit more often than others do because I jump into building with not much planning beforehand. I guess we are a different kind of website builders when it comes to that.

> Putting the entire configuration of an image style in a dialog would be challenging. There are just too many steps involved in making an image style.

I do not want to have the entire config done in one go from the add image field form. I simply want to be able to select a new image style (other than the defaults or the already configured custom ones) from the drop-down menu. I don't mind tweaking the style afterwards. My main concern is the bad UX of having to re-do the entire add image field form just because I forgot to create my image styles beforehand.

How about this then: Could we perhaps add an extra "dynamic" item as an option in the drop-down (something like "create new style")*. This specific image style would be created as a "placeholder" (with no effects) and would be associated with the image field being edited/created. Once the image field form is submitted, instead of being redirected to the manage fields page of the content type, the user is taken to the configure/edit image style for the placeholder image style they just created where they can add effects at will. Once they submit that form, then they are finally redirected to the content type manage fields page.

*If the user selects that, then the only thing that we ask as input is the image style name (and its machine name of course). We can show these fields either inline using #states, or if you think that "cannibalizing" one form to insert a some elements from another is not acceptable, then we can ask for these upon form submit as a second, wizard-like step. Point is to not have the user hop from page to page in the admin UI in order to accomplish what to them seems like a single task (it is actually two combined in one: adding an image field + an image style).

**Comment by klonos** [comment]
> The only way I can think of making this possible would be if we first made the configuration of image styles an entirely one-page process (with lots of AJAX).

I meant to file a separate issue for that :smile:

**Comment by Graham-72** [comment]
I like the idea of using a placeholder, rather as when adding a custom style to a field in a view I give the style a name before defining all aspects in CSS. I think a creative design process needs to be able to react to appearance on a screen and not have to work it all out in advance, so I am often saying to myself (or my co-worker) 'I think we will try that text a little bit larger', simlarly 'I think we will try that thumbnail a little bit larger'. But is it a big deal compared with other things we could improve?

**Comment by mikemccaffrey** [comment]
> I think that I work more like @Graham-72 too and so I get to this situation a bit more often than others do because I jump into building with not much planning beforehand. I guess we are a different kind of website builders when it comes to that, but it comes with building small sites.

Well, that is a good perspective to consider, and I do think that there are opportunities to improve the interface around creating, managing, and using image styles, so it definitely warrants a discussion.

> But is it a big deal compared with other things we could improve?

However, I too am thinking that this is a fairly small usability issue compared to some of our other challenges, and perhaps not the best use of our limited time before the next release. I also worry that a change to address this issue could add some undesired complexity to both the code and the workflow if not implemented in the right way.

Maybe we can leave this issue open and see if we can pull it into some larger discussion about either image, field, or form UX some time in the future after the next release?

**Comment by gifad** [comment]
> But is it a big deal compared with other things we could improve?

For instance, make an image style applied in CKEditor image plugin…
That's far easier than in drupal 7, since backdrop knows the file id of the image.

And this would bring backdrop closer to feature parity with drupal 6 (yes, 6, this is not a typo)

**Comment by klonos** [comment]
> However, I too am thinking that this is a fairly small usability issue compared to some of our other challenges, and perhaps not the best use of our limited time before the next release.

No milestone was set for this. If I see something that needs to fixed or improved, I file an issue for it (when I have the time to do so). I see the issue queue as a big to-do list that will never be "empty". I do not nag or "push" for things to happen. Some day, the small issues, the ones that can be considered low-hanging fruits, might be picked up by people that slowly make their "jump" from being site builders to module developers. That might even be me someday. So I like to have a place where all my ideas are logged. I hope this is not perceived as "noise".

Out of the thousands of issues, people can search, filter and choose the ones to work on. Later on, perhaps we might even have something like https://github.com/backdrop-ops/backdropcms.org/issues/211 to help us pick the important ones :wink:

**Comment by Graham-72** [comment]
@klonos I fully understand and very much appreciate your approach to finding these things and creating issues, and no, you doing this is not noise. I was only trying to express an opinion about whether it was a major issue for me because I keep on finding more than enough issues to deal with myself, and reading about many many more. So I was in part trying to restrain myself from spending more time on this right now. No criticism of yourself or others intended. On the contrary, respect for all you do.

**Comment by gifad** [comment]
> > But is it a big deal compared with other things we could improve?

>For instance, make an image style applied in CKEditor image plugin…

Quick and dirty patch to [CKEditor_addon](https://github.com/backdrop-contrib/ckeditor_addon)
visible at http://p1347.phpnet.org/backdrop/l/3rgz8UIEd0Vd/
![capture d ecran 2016-03-30 a 00 48 03](https://cloud.githubusercontent.com/assets/15943462/14126437/4609d5f0-f611-11e5-8a2b-36c32e84b7a9.jpg)

I'm sorry, @klonos, to distort your issue, but I like better work for site contributors than site "builders"

**Comment by Graham-72** [comment]
> I like better work for site contributors than site "builders"

Moi aussi
:+1: for the quick and dirty patch
But I am also very grateful for those working hard to improve core.

**Comment by mikemccaffrey** [comment]
> @klonos: No milestone was set for this. If I see something that needs to be fixed or improved, I file an issue for it (when I have the time to do so). I see the issue queue as a big to-do list that will never be "empty". I do not nag or "push" for things to happen. Some day, the small issues, the ones that can be considered low-hanging fruits, might be picked up by people that slowly make their "jump" from being site builders to module developers. That might even be me someday. So I like to have a place where all my ideas are logged. I hope this is not perceived as "noise".

Don't worry, the more tickets the better! I too post plenty of random ideas as they come to my mind. 

And I'm not saying that you or anyone else shouldn't work on the ticket, if you are so inclined. However, there actually was a 1.4.0 milestone set on the issue, and I was just explaining that I was removing it, since it didn't seem like an issue we would need to resolve before the release next month.

**Comment by klonos** [comment]
Fair enough @mikemccaffrey I understand and agree.
